### PR TITLE
Bring back conventions on configurable file collections

### PR DESCRIPTION
**Pull request by abstratt**
Fixes: #21752 and #25509

Reverts: #28135

### CODE REVIEW

**Comment by abstratt** [comment]
@bot-gradle merge

**Comment by abstratt** [comment]
@lkasso The bulk of this PR is re-enabling/exposing functionality that was disabled/hidden just before RC1 to unblock the release. The only new content is the release note addition, appreciate having your 👀 .

File: platforms/documentation/docs/src/docs/release/notes.md [hunk 1]
```diff
@@ -52,7 +52,36 @@ Example:
 ADD RELEASE FEATURES BELOW
 vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv -->
 
+#### Ability to set conventions on file collections
 
+Plugin-provided tasks often expose file collections that are meant to be customizable by build engineers (for instance, the classpath for the JavaCompile task).
+Up to now, in order for plugin authors to define default values for file collections, they had to resort to configuring those defaults as initial values.
```
Comment (lkasso) [inline platforms/documentation/docs/src/docs/release/notes.md:58]:
```suggestion
Up until now, for plugin authors to define default values for file collections, they have had to resort to configuring those defaults as initial values.
```

File: platforms/documentation/docs/src/docs/release/notes.md [hunk 2]
```diff
@@ -52,7 +52,36 @@ Example:
...
+Conventions provide a better model for that: plugin authors recommend default values via conventions, and users choose to accept, add on top, or completely 
+replace them when defining their actual value.
+
+This release introduces a  pair of [`convention(...)`](javadoc/org/gradle/api/file/ConfigurableFileCollection.html#convention-java.lang.Object...-) methods 
+on `ConfigurableFileCollection` that define the default value of a file collection, if no explicit value is previously set via `setFrom(...)` or `from(...)`.
```
Comment (lkasso) [inline platforms/documentation/docs/src/docs/release/notes.md:63]:
```suggestion
on `ConfigurableFileCollection` that define the default value of a file collection if no explicit value is previously set via `setFrom(...)` or `from(...)`.
```

File: platforms/documentation/docs/src/docs/release/notes.md [hunk 3]
```diff
@@ -52,7 +52,36 @@ Example:
...
+
+```kotlin
+val files = objects.fileCollection().convention("dir1")
+files.from("dir2")
+
+println(files.elements.get()) // [.../dir1, .../dir2]
+```
+
+`#from(...)` will honor the convention if one is configured at the time it is invoked, so the order of operations will matter.
```
Comment (lkasso) [inline platforms/documentation/docs/src/docs/release/notes.md:72]:
```suggestion
`#from(...)` will honor the convention if one is configured when invoked, so the order of operations will matter. 
```

File: platforms/documentation/docs/src/docs/release/notes.md [hunk 4]
```diff
@@ -52,7 +52,36 @@ Example:
...
+```kotlin
+val files = objects.fileCollection().convention("dir1")
+files.from("dir2")
+
+println(files.elements.get()) // [.../dir1, .../dir2]
+```
+
+`#from(...)` will honor the convention if one is configured at the time it is invoked, so the order of operations will matter. 
+
+In order to forcefully override or prevent a convention (i.e., regardless the order of those operations), one should use `#setFrom()` instead:
```
Comment (lkasso) [inline platforms/documentation/docs/src/docs/release/notes.md:74]:
```suggestion
To forcefully override or prevent a convention (i.e., regardless of the order of those operations), one should use `#setFrom()` instead:
```

**Review by lkasso — APPROVED** [review]
Some recommended changes for release notes.

**Comment by abstratt** [comment]
@bot-gradle merge

**Pull request merged**